lib/commands/exam.js

```diff
@@ -120,141 +124,97 @@ module.exports = TestCommand.extend({
+   * the number of launchers to run in parallel in their testem.json config.
    *
+   * @param {Object} commandOptions
    * @override
    */
   _generateCustomConfigs(commandOptions) {
     const config = this._super._generateCustomConfigs.apply(this, arguments);
 
-    if (this.validator.shouldParallelize) {
-      // Get the testPage from the generated config or the Testem config and
-      // use it as the baseUrl for the parallelized test pages
-      const baseUrl = config.testPage || this._getTestPage(commandOptions.configFile);
-      const count = commandOptions.split;
-      let partitions = commandOptions.partition;
-
-      if (!partitions) {
-        partitions = [];
-        for (let i = 0; i < commandOptions.split; i++) {
-          partitions.push(i + 1);
-        }
-      }
+    if (!commandOptions.loadBalance && !commandOptions.replayExecution && !commandOptions.parallel) return config;
 
-      if (Array.isArray(baseUrl)) {
-        const command = this;
-        config.testPage = baseUrl.reduce(function(testPages, baseUrl) {
-          return testPages.concat(command._generateTestPages(baseUrl, partitions, count));
-        }, []);
-      } else {
-        config.testPage = this._generateTestPages(baseUrl, partitions, count);
+    config.testPage = getMultipleTestPages(config, commandOptions);
+
+    if (commandOptions.loadBalance || commandOptions.replayExecution) {
+      config.custom_browser_socket_events = this._addLoadBalancingBrowserSocketEvents(config);
+      if (commandOptions.replayExecution) {
+        this.testemEvents.setReplayExecutionMap(commandOptions.replayExecution, commandOptions.replayBrowser);
       }
     }
 
     return config;
   },
 
   /**
-   * Gets the test page specified by the application's Testem config.
+   * Adds additional event handlers which enables load balancing.
    *
-   * @param {String} [configFile] - Path to the config file to use
-   * @return {String} testPage
+   * @param {Object} config
+   * @return {Object} events
    */
-  _getTestPage(configFile) {
-    // Attempt to read in the testem config and use the test_page definition
-    const testemConfig = this._readTestemConfig(configFile);
-    let testPage = testemConfig && testemConfig.test_page;
-
-    // If there is no test_page to use as the testPage, we warn that we're using
-    // a default value
-    if (!testPage) {
-      // eslint-disable-next-line no-console
-      console.warn('No test_page value found in the config. Defaulting to "tests/index.html?hidepassed"');
-      testPage = 'tests/index.html?hidepassed';
+  _addLoadBalancingBrowserSocketEvents(config) {
+    const commands = this.commands;
+    const testemEvents = this.testemEvents;
+    const ui = this.ui;
+    const events = config.custom_browser_socket_events || {};
+    const testExecutionFileName = `test-execution-${Date.now()}.json`;
+    const getBrowserId = function(testPage) {
```

Comment:
The nested function is in `_addLoadBalancingBrowserSocketEvents` because we want the function to be accessible. `_addLoadBalancingBrowserSocketEvents`  is executed in testem, not in ember-exam `getBrowserId` should be within the function.